datafusion/core/src/execution/memory_manager/mod.rs

```diff
@@ -97,339 +89,154 @@ impl MemoryManagerConfig {
-        );
-        assert!(*requesters_total >= freed);
-        *requesters_total -= freed;
-        *requesters_total += acquired;
-        self.cv.notify_all();
+    /// Frees all bytes from this allocation returning the number of bytes freed
+    pub fn free(&mut self) -> usize {
+        let size = self.size;
+        if size != 0 {
+            self.shrink(size)
+        }
+        size
+    }
+
+    /// Frees `capacity` bytes from this allocation
+    ///
+    /// # Panics
+    ///
+    /// Panics if `capacity` exceeds [`Self::size`]
+    pub fn shrink(&mut self, capacity: usize) {
+        let new_size = self.size.checked_sub(capacity).unwrap();
+        self.state.used.fetch_sub(capacity, Ordering::SeqCst);
+        self.size = new_size
+    }
+
+    /// Sets the size of this allocation to `capacity`
+    pub fn resize(&mut self, capacity: usize) {
+        use std::cmp::Ordering;
+        match capacity.cmp(&self.size) {
+            Ordering::Greater => self.grow(capacity - self.size),
+            Ordering::Less => self.shrink(self.size - capacity),
+            _ => {}
+        }
     }
 
-    fn record_free(&self, freed: usize) {
-        let mut requesters_total = self.requesters_total.lock();
-        debug!(
-            "free: total {}, freed {}",
-            human_readable_size(*requesters_total),
-            human_readable_size(freed)
-        );
-        assert!(*requesters_total >= freed);
-        *requesters_total -= freed;
-        self.cv.notify_all();
+    /// Increase the size of this by `capacity` bytes
+    pub fn grow(&mut self, capacity: usize) {
+        self.state.used.fetch_add(capacity, Ordering::SeqCst);
+        self.size += capacity;
     }
 
-    /// Drop a memory consumer and reclaim the memory
-    pub(crate) fn drop_consumer(&self, id: &MemoryConsumerId, mem_used: usize) {
-        // find in requesters first
-        {
-            let mut requesters = self.requesters.lock();
-            if requesters.remove(id) {
-                let mut total = self.requesters_total.lock();
-                assert!(*total >= mem_used);
-                *total -= mem_used;
-                self.cv.notify_all();
-                return;
-            }
-        }
-        self.shrink_tracker_usage(mem_used);
-        self.cv.notify_all();
+    /// Try to increase the size of this [`TrackedAllocation`] by `capacity` bytes
+    pub fn try_grow(&mut self, capacity: usize) -> Result<()> {
+        let pool_size = self.state.pool_size;
+        self.state
+            .used
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |used| {
+                let new_used = used + capacity;
+                (new_used <= pool_size).then_some(new_used)
+            })
+            .map_err(|used| DataFusionError::ResourcesExhausted(format!("Failed to allocate additional {} bytes for {} with {} bytes already allocated - maximum available is {}", capacity, self.name, self.size, used)))?;
```

Comment:
I think the order of the parameters is incorrect but I may be misreading it

```suggestion
            .map_err(|used| DataFusionError::ResourcesExhausted(format!("Failed to allocate additional {} bytes for {} with {} bytes already allocated - maximum available is {}", capacity, self.name, used, self.size)))?;
```